tools/imatrix/imatrix.cpp

```diff
@@ -125,95 +128,241 @@ static void process_tensor_name(const std::string & input, std::string & layer,
-    if (act_dev > 0.0f) {
+    int zd_score = 0;
+    if (std_deviation > 0.0f) {
         for (const auto act : activations) {
-            if (const float p = (act - act_mean) / act_dev; p > 1) {
-                z_score++;
-            }
+            if (const float z = (act - mean) / std_deviation; std::fabs(z) > 1.0f) zd_score++;
         }
     }
 
     auto & ts = tstats.emplace_back();
-    ts.tensor     = name;
-    ts.stats      = e;
-    ts.total_sqract = act_total;
-    ts.mean_sqract  = act_mean;
-    ts.max_sqract   = act_max;
-    ts.min_sqract   = act_min;
-    ts.elements   = static_cast<int>(activations.size());
-    ts.stddev     = act_dev;
-    ts.active     = active_ratio;
-    ts.entropy    = entropy;
-    ts.zd         = static_cast<float>(z_score) / ts.elements;
+    ts.tensor       = name;
+    ts.stats        = e;
+    ts.sum_values   = sum;
+    ts.mean_values  = mean;
+    ts.max_values   = max;
+    ts.min_values   = min;
+    ts.elements     = static_cast<int>(activations.size());
+    ts.std_deviation       = std_deviation;
+    ts.entropy      = entropy;
+    ts.zd_score     = static_cast<float>(zd_score) / ts.elements;
+
+    return e.activations.empty();
 }
 
-static void compute_cossim(std::vector<tensor_statistics> & tstats) {
+static void compute_tensor_statistics(std::vector<tensor_statistics> & tstats) {
     static const std::regex pattern(R"(blk\.(\d+)\.)");
+
+    // compute the Cosine Similarity between the same tensors in consecutive layers
     for (auto & ts : tstats) {
+        ts.cossim = 0;
+
         if (std::smatch match; std::regex_search(ts.tensor, match, pattern)) {
             const int blk = std::stoi(match[1]);
+            if (blk <= 0) continue;
             std::string tname(ts.tensor);
             tname.replace(match.position(1), match.length(1), std::to_string(blk-1));
             auto prev = std::find_if(tstats.begin(), tstats.end(),
                 [tname](const tensor_statistics & t) { return t.tensor == tname; });
-            if (prev != tstats.end()) {
-                const float dp = std::inner_product(ts.stats.values.begin(), ts.stats.values.end(),
-                    prev->stats.values.begin(), 0.0f);
-                const float curr_mag = std::sqrt(std::inner_product(ts.stats.values.begin(), ts.stats.values.end(),
-                    ts.stats.values.begin(), 0.0f));
-                const float prev_mag = std::sqrt(std::inner_product(prev->stats.values.begin(), prev->stats.values.end(),
-                    prev->stats.values.begin(), 0.0f));
-                const float cs = dp / (curr_mag * prev_mag);
-                ts.cossim = cs;
+            if (prev == tstats.end()) continue;
+            const auto curr_avg = compute_tensor_averages(ts.stats);
+            const auto prev_avg = compute_tensor_averages(prev->stats);
+            if (curr_avg.size() == prev_avg.size() && !curr_avg.empty()) {
+                float dot_prod = 0.0f, vec1 = 0.0f, vec2 = 0.0f;
```

Comment:
(style) Consider using one line per variable definition.

```suggestion
                float dot_prod = 0.0f;
                float vec1 = 0.0f;
                float vec2 = 0.0f;
```